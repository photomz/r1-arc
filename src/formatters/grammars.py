from lark import Lark, UnexpectedToken, Tree
from lark.indenter import PythonIndenter
from src.utils import ROOT

# Load Python grammar from Lark
kwargs = dict(postlex=PythonIndenter(), parser="lalr")

larkf = ROOT / "src/formatters/python.lark"
lark_schema = larkf.open().readlines()

pydsl_parser = Lark.open(larkf, start="start", **kwargs)
parse = lambda s, **kw: pydsl_parser.parse(s, on_error=handle_error, **kw)
# python3_parser = Lark.open_from_package(
#     "lark", "python.lark", ["grammars"], start="file_input", **kwargs
# )


class StreamingParser:
    def __init__(self, parser):
        self.parser = parser

    def parse(self, text):
        try:
            return self.parser.parse(text)
        except UnexpectedToken as e:
            # if e.token.type == "$END":
            # Get the last valid tree from the stack
            for item in reversed(e.state.value_stack):
                if isinstance(item, (Tree)):
                    return item
            # If no tree found, create a minimal valid tree
            return Tree("start", [])
            # raise


parser = StreamingParser(pydsl_parser)

# md disable: blockquotes
completion = """
<think>
Hello I'm thinking about ... a hash # 

Wait, Case 2 ...

This is **bold**
With _italics_ and code
* Lists
# Headers
[Links](https://example.com)

Still case 2 ...

Alternative, ... take 3

I'm stuck here.... case 4

</think>
# Solution

My solution is blah blah ... 
```py
def solve(I):
    x1 = partition(I)
    x2 = argmin(x1, size)
    O = subgrid(x2, I)
    return O
```

Explanation of DSL: ....
""".strip()

tree = pydsl_parser.parse(completion)
print(tree.pretty())

dsl_code = """
def solve(I):
    x1 = partition(I)
    x2 = argmin(x1, size)
    O = subgrid(x2, I)
    return O
"""

# tree = python3_parser.parse(dsl_code)
# print(tree.pretty())
