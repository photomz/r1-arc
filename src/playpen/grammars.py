from lark import Lark
from lark.indenter import PythonIndenter

# Load Python grammar from Lark
kwargs = dict(postlex=PythonIndenter(), start="start")

# Official Python grammar by Lark
python_parser3 = Lark.open("./src/playpen/python.lark", parser="lalr", **kwargs)

# dsl_grammar = """
# %import python.lark

# // Restrict function calls to only allow DSL function names
# %override atom_expr: DSL_FUNC "(" arguments? ")"

# // Define the set of valid DSL functions
# DSL_FUNC: "valmin" | "argmax" | "argmin" | "hmirror" | "subgrid" | "partition" | "recolor" | "move" | "ofcolor"

# // Restrict valid statements to function definitions, assignments, and return statements
# stmt: funcdef | assignment | return_stmt

# // Keep expressions but only allow simple function calls or variables
# expr: funccall | VAR
# """

# dsl_parser = Lark(dsl_grammar, parser="lalr", **kwargs)


# Example DSL code snippet
dsl_code = """
```py
def solve(I):
    x1 = partition(I)
    x2 = argmin(x1, size)
    O = subgrid(x2, I)
    return O
```
"""

tree = python_parser3.parse(dsl_code)
print(tree.pretty())
