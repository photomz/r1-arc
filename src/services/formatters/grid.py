from abc import ABC, abstractmethod
from dataclasses import dataclass
from collections import defaultdict
from enum import Enum
from typing import TypeVar
from itertools import product
import numpy as np
import typer
from pathlib import Path
import json
from devtools import debug

from src.utils.devtools import debug

# (A - Z) + (AA - AD)
ALPHA = list(chr(ord("A") + i) for i in range(26)) + list(
    f"A{chr(ord('A') + i)}" for i in range(4)
)

Grid = TypeVar("Grid", bound=np.ndarray)


def bucket_diffs(g1: Grid, g2: Grid) -> dict[tuple[int, int], list[tuple[int, int]]]:
    """Bucket diffs by index (color)"""
    x, y = (g1 != g2).nonzero()
    diffs = defaultdict(list)

    for i, j in zip(x, y, strict=True):
        pair = (g1[i, j], g2[i, j])
        diffs[pair].append((int(i), int(j)))

    return diffs


@dataclass
class Formatter(ABC):
    sep: str = "|"
    newline: str = "\n"
    n_max: int = 30

    @classmethod
    def format(cls, grid: Grid) -> str:
        """Base formatter"""
        if grid.shape[0] > cls.n_max or grid.shape[1] > cls.n_max:
            raise ValueError(f"Grid dimensions exceed maximum size of {cls.n_max}")
        return ""


@dataclass
class GridFormatter(Formatter):
    """
    0 | 0 | 0
    0 | 1 | 0
    0 | 0 | 0
    """

    @classmethod
    def format(cls, grid: Grid) -> str:
        super().format(grid)
        return cls.newline.join(cls.sep.join(str(x) for x in row) for row in grid)


@dataclass
class SpreadsheetFormatter(Formatter):
    """
        A | B | C
    1	0 | 0 | 0
    2	0 | 1 | 0
    3 	0 | 0 | 0
    """

    @classmethod
    def format(cls, grid: Grid) -> str:
        super().format(grid)
        _, n_cols = grid.shape

        header = "\t" + cls.sep.join([""] + ALPHA[:n_cols])
        rows = "\n".join(
            str(i + 1) + "\t" + cls.sep.join([str(x) for x in row])
            for i, row in enumerate(grid)
        )
        return cls.newline.join([header, rows])


@dataclass
class Differ(Formatter):
    change: str = " -> "
    blank: str = "  --  "

    @classmethod
    def diff(self, A: Grid, B: Grid) -> str:
        """Base diff 2 grids"""
        super().format(A)
        super().format(B)

        if A.shape != B.shape:
            raise ValueError(f"Unexpected shape: A {A.shape}, B {B.shape}")
        return ""


@dataclass
class ListDiffer(Differ):

    @classmethod
    def diff(self, A: Grid, B: Grid) -> str:
        """List A->B color diffs like adjacency list"""
        super().diff(A, B)

        diffs = bucket_diffs(A, B)
        output = []

        for (cA, cB), coords in sorted(diffs.items()):
            cells = " ".join(f"{ALPHA[i]}{j+1}" for i, j in sorted(coords))
            output.append(f"{cA}{self.change}{cB}: {cells}")

        return "\n".join(output)


@dataclass
class CellDiffer(Differ):

    @classmethod
    def diff(self, A: Grid, B: Grid) -> str:
        """Normal grid, but each cell is (ij)->(ij)' diff"""
        super().diff(A, B)

        rows, cols = A.shape
        # Init with blank cells
        diff_grid = [[self.blank for _ in range(cols)] for _ in range(rows)]

        for i, j in product(range(rows), range(cols)):
            if A[i, j] != B[i, j]:
                diff_grid[i][j] = f"{A[i,j]}{self.change}{B[i,j]}"

        make_row = lambda row: self.sep.join(str(x) for x in row)

        return self.newline.join(map(make_row, diff_grid))


app = typer.Typer()


class FormatterNames(str, Enum):
    GRID = "grid"
    SPREADSHEET = "spreadsheet"
    LIST_DIFF = "diff-list"
    CELL_DIFF = "diff-cell"


TASK_PATH = Path(__file__).parent.parent.parent / "tasks"
FORMATTERS = {
    FormatterNames.GRID: GridFormatter,
    FormatterNames.SPREADSHEET: SpreadsheetFormatter,
    FormatterNames.LIST_DIFF: ListDiffer,
    FormatterNames.CELL_DIFF: CellDiffer,
}


@app.command()
def format(
    id: str, filename: str = "train.json", type: FormatterNames = FormatterNames.GRID
):
    """Format or diff all grids in ARC task"""
    with open(TASK_PATH / filename) as f:
        task = json.load(f)[id]

    formatter = FORMATTERS[type]
    is_diff = type in {FormatterNames.LIST_DIFF, FormatterNames.CELL_DIFF}

    grids = task["train"] + task["test"]

    for i, g in enumerate(grids, 1):
        print(f"> Grid {i if i < len(grids) else 'TEST'}")

        A = np.array(g["input"])
        B = np.array(g.get("output", []))

        # B.size skips held-out test grid output
        if is_diff:
            if B.size:
                diff = formatter.diff(A, B)
                debug(diff)
        else:
            finput = formatter.format(A)
            debug(finput)

            if B.size:
                foutput = formatter.format(B)
                debug(foutput)


if __name__ == "__main__":
    app()
