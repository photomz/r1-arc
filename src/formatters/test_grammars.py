from deepseek_tokenizer import ds_token
from src.formatters.grammars import parser

completion = """
<think>
Hello I'm thinking about ... a hash # 

Wait, Case 2 ...

This is **bold**
With _italics_ and code
* Lists
# Headers
[Links](https://example.com)

Still case 2 ...

Alternative, ... take 3

I'm stuck here.... case 4

</think>
# Solution

My solution is blah blah ... 
```py
def solve(I):
    x1 = partition(I)
    x2 = argmin(x1, size)
    O = subgrid(x2, I)
    return O
```

Explanation of DSL: ....
""".strip()


def test_streaming(snapshot):
    token_ids = ds_token.encode(completion)
    token_stream = [ds_token.decode(token_ids[:i]) for i in range(len(token_ids))]

    parses = []
    for pre in token_stream:
        tree = parser.parse(pre)
        parses.append(tree.pretty() if tree else "No parse")
    assert parses == snapshot
