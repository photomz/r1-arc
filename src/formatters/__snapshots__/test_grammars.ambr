# serializer version: 1
# name: test_streaming
  list([
    '''
      start
  
    ''',
    '''
      start
  
    ''',
    '''
      start
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __start_star_0
  
    ''',
    '''
      __think_star_1	Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
  
    ''',
    '''
      __think_star_1	Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      think
        <think>
        Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
        </think>
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      __file_input_star_2
  
    ''',
    '''
      var
        name	x1
  
    ''',
    '''
      var
        name	x1
  
    ''',
    '''
      name	partition
  
    ''',
    '''
      arguments
        var
          name	I
  
    ''',
    '''
      arguments
        var
          name	I
  
    ''',
    '''
      assign_stmt
        assign
          var
            name	x1
          funccall
            name	partition
            arguments
              var
                name	I
  
    ''',
    '''
      __suite_plus_24
        assign_stmt
          assign
            var
              name	x1
            funccall
              name	partition
              arguments
                var
                  name	I
  
    ''',
    '''
      var
        name	x2
  
    ''',
    '''
      var
        name	x2
  
    ''',
    '''
      var
        name	x2
  
    ''',
    '''
      name	argmin
  
    ''',
    '''
      name	argmin
  
    ''',
    '''
      var
        name	x1
  
    ''',
    '''
      var
        name	x1
  
    ''',
    '''
      arguments
        var
          name	x1
        var
          name	size
  
    ''',
    '''
      arguments
        var
          name	x1
        var
          name	size
  
    ''',
    '''
      __suite_plus_24
        assign_stmt
          assign
            var
              name	x1
            funccall
              name	partition
              arguments
                var
                  name	I
        assign_stmt
          assign
            var
              name	x2
            funccall
              name	argmin
              arguments
                var
                  name	x1
                var
                  name	size
  
    ''',
    '''
      var
        name	O
  
    ''',
    '''
      var
        name	O
  
    ''',
    '''
      var
        name	O
  
    ''',
    '''
      name	subgrid
  
    ''',
    '''
      name	subgrid
  
    ''',
    '''
      var
        name	x2
  
    ''',
    '''
      var
        name	x2
  
    ''',
    '''
      arguments
        var
          name	x2
        var
          name	I
  
    ''',
    '''
      arguments
        var
          name	x2
        var
          name	I
  
    ''',
    '''
      __suite_plus_24
        assign_stmt
          assign
            var
              name	x1
            funccall
              name	partition
              arguments
                var
                  name	I
        assign_stmt
          assign
            var
              name	x2
            funccall
              name	argmin
              arguments
                var
                  name	x1
                var
                  name	size
        assign_stmt
          assign
            var
              name	O
            funccall
              name	subgrid
              arguments
                var
                  name	x2
                var
                  name	I
  
    ''',
    '''
      __suite_plus_24
        assign_stmt
          assign
            var
              name	x1
            funccall
              name	partition
              arguments
                var
                  name	I
        assign_stmt
          assign
            var
              name	x2
            funccall
              name	argmin
              arguments
                var
                  name	x1
                var
                  name	size
        assign_stmt
          assign
            var
              name	O
            funccall
              name	subgrid
              arguments
                var
                  name	x2
                var
                  name	I
  
    ''',
    '''
      __suite_plus_24
        assign_stmt
          assign
            var
              name	x1
            funccall
              name	partition
              arguments
                var
                  name	I
        assign_stmt
          assign
            var
              name	x2
            funccall
              name	argmin
              arguments
                var
                  name	x1
                var
                  name	size
        assign_stmt
          assign
            var
              name	O
            funccall
              name	subgrid
              arguments
                var
                  name	x2
                var
                  name	I
  
    ''',
    '''
      start
        think
          <think>
          Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
          </think>
        answer
          
      # Solution
      
      My solution is blah blah ... 
      
          pyblock
            ```py
            funcdef
              None
              suite
                assign_stmt
                  assign
                    var
                      name	x1
                    funccall
                      name	partition
                      arguments
                        var
                          name	I
                assign_stmt
                  assign
                    var
                      name	x2
                    funccall
                      name	argmin
                      arguments
                        var
                          name	x1
                        var
                          name	size
                assign_stmt
                  assign
                    var
                      name	O
                    funccall
                      name	subgrid
                      arguments
                        var
                          name	x2
                        var
                          name	I
                return_stmt
                  var
                    name	O
            ```
  
    ''',
    '''
      start
        think
          <think>
          Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
          </think>
        answer
          
      # Solution
      
      My solution is blah blah ... 
      
          pyblock
            ```py
            funcdef
              None
              suite
                assign_stmt
                  assign
                    var
                      name	x1
                    funccall
                      name	partition
                      arguments
                        var
                          name	I
                assign_stmt
                  assign
                    var
                      name	x2
                    funccall
                      name	argmin
                      arguments
                        var
                          name	x1
                        var
                          name	size
                assign_stmt
                  assign
                    var
                      name	O
                    funccall
                      name	subgrid
                      arguments
                        var
                          name	x2
                        var
                          name	I
                return_stmt
                  var
                    name	O
            ```
          
      
      Explanation
  
    ''',
    '''
      start
        think
          <think>
          Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
          </think>
        answer
          
      # Solution
      
      My solution is blah blah ... 
      
          pyblock
            ```py
            funcdef
              None
              suite
                assign_stmt
                  assign
                    var
                      name	x1
                    funccall
                      name	partition
                      arguments
                        var
                          name	I
                assign_stmt
                  assign
                    var
                      name	x2
                    funccall
                      name	argmin
                      arguments
                        var
                          name	x1
                        var
                          name	size
                assign_stmt
                  assign
                    var
                      name	O
                    funccall
                      name	subgrid
                      arguments
                        var
                          name	x2
                        var
                          name	I
                return_stmt
                  var
                    name	O
            ```
          
      
      Explanation of
  
    ''',
    '''
      start
        think
          <think>
          Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
          </think>
        answer
          
      # Solution
      
      My solution is blah blah ... 
      
          pyblock
            ```py
            funcdef
              None
              suite
                assign_stmt
                  assign
                    var
                      name	x1
                    funccall
                      name	partition
                      arguments
                        var
                          name	I
                assign_stmt
                  assign
                    var
                      name	x2
                    funccall
                      name	argmin
                      arguments
                        var
                          name	x1
                        var
                          name	size
                assign_stmt
                  assign
                    var
                      name	O
                    funccall
                      name	subgrid
                      arguments
                        var
                          name	x2
                        var
                          name	I
                return_stmt
                  var
                    name	O
            ```
          
      
      Explanation of DSL
  
    ''',
    '''
      start
        think
          <think>
          Hello I'm thinking about ... a hash # 
      
      Wait, Case 2 ...
      
      This is **bold**
      With _italics_ and code
      * Lists
      # Headers
      [Links](https://example.com)
      
      Still case 2 ...
      
      Alternative, ... take 3
      
      I'm stuck here.... case 4
      
      
          </think>
        answer
          
      # Solution
      
      My solution is blah blah ... 
      
          pyblock
            ```py
            funcdef
              None
              suite
                assign_stmt
                  assign
                    var
                      name	x1
                    funccall
                      name	partition
                      arguments
                        var
                          name	I
                assign_stmt
                  assign
                    var
                      name	x2
                    funccall
                      name	argmin
                      arguments
                        var
                          name	x1
                        var
                          name	size
                assign_stmt
                  assign
                    var
                      name	O
                    funccall
                      name	subgrid
                      arguments
                        var
                          name	x2
                        var
                          name	I
                return_stmt
                  var
                    name	O
            ```
          
      
      Explanation of DSL:
  
    ''',
  ])
# ---
